### PR TITLE
Shahryart/typechecking enummap

### PR DESCRIPTION
**Pull request by shahryart**
typecheck enum map

### CODE REVIEW

File: graphql/schemabuilder/reflect.go [hunk 1]
```diff
@@ -184,6 +185,8 @@ func (sb *schemaBuilder) getEnumArgParser(typ reflect.Type) (*argParser, graphql
 		if !ok {
 			return fmt.Errorf("unknown enum value %v", asString)
 		}
+		log.Println(dest.Type())
+		log.Println(reflect.TypeOf(val))
```
Comment (stephen) [inline graphql/schemabuilder/reflect.go]: remove these

File: graphql/schemabuilder/reflect.go [hunk 2]
```diff
@@ -818,19 +821,23 @@ func (s *Schema) Enum(val interface{}, enumMap interface{}) {
...
+			if val.Type().Kind() != typ.Kind() {
+				panic("types are not equal")
+			}
 			if key.Kind() == reflect.String {
-				eMap[key.String()] = val.Interface()
+				mapVal := val.Convert(typ)
+				eMap[key.String()] = mapVal.Interface()
 			} else {
 				panic("keys are not strings")
 			}
```
Comment (stephen) [inline graphql/schemabuilder/reflect.go]: can we change the panic below to be lowercased

File: graphql/schemabuilder/reflect_test.go [hunk 3]
```diff
@@ -210,20 +226,65 @@ func TestEnumMapWrongArg(t *testing.T) {
 
 }
 
-func TestEnumMapKeys(t *testing.T) {
+func TestEnumMapWrongArg2(t *testing.T) {
```
Comment (stephen) [inline graphql/schemabuilder/reflect_test.go]: Let's make the test names more descriptive to the case, i.e.`TestEnumMapMismatchedValues`

File: graphql/schemabuilder/reflect_test.go [hunk 4]
```diff
@@ -210,20 +226,65 @@ func TestEnumMapWrongArg(t *testing.T) {
...
 	}()
-	type enumType int32
-	schema.Enum(enumType(1), map[int32]int32{
-		1: 1,
+
+	schema.Enum(enumType3(1), map[string]interface{}{
+		"firstField":  interface{}(1),
+		"secondField": interface{}(2),
+		"thirdField":  interface{}(3),
 	})
```
Comment (stephen) [inline graphql/schemabuilder/reflect_test.go]: i think there are ways we could support this case: https://play.golang.org/p/9bjfb8w_8h2

File: graphql/schemabuilder/reflect_test.go [hunk 5]
```diff
@@ -196,6 +196,22 @@ func TestExecuteGood(t *testing.T) {
 	}
 }
 
+func TestEnumMapKeys(t *testing.T) {
+	schema := NewSchema()
+	defer func() {
+		r := recover()
+		if r == nil {
```
Comment (stephen) [inline graphql/schemabuilder/reflect_test.go]: let's assert on specific error messages here

File: graphql/schemabuilder/reflect_test.go (hunk 4 above)
Comment (shahryart) [reply graphql/schemabuilder/reflect_test.go]: It doesn't differentiate between int32 and int

Comment (stephen) [reply graphql/schemabuilder/reflect_test.go]: could you paste a code example? https://play.golang.org/p/ItVllolNiF2

Comment (shahryart) [reply graphql/schemabuilder/reflect_test.go]: This particular test case will still fail because the integers aren't of int32 type

Comment (stephen) [reply graphql/schemabuilder/reflect_test.go]: ah, that makes sense.

**Pull request merged**